File: examples/python/test_connection.py
#!/usr/bin/env python
# TWiT API Connection Test Script
# This script tests the connection to the TWiT.tv API

import requests
import json
import yaml
import os
import sys
from pathlib import Path

# Load credentials from shared credentials file
script_dir = os.path.dirname(os.path.abspath(__file__))
credentials_file = os.path.join(script_dir, '..', 'credentials.yml')

if os.path.exists(credentials_file):
    try:
        print(f"Loading credentials from: {credentials_file}")
        with open(credentials_file, 'r') as file:
            credentials_content = file.read()
            print(f"Credentials file content (first line): {credentials_content.splitlines()[0].strip()}")

        with open(credentials_file, 'r') as file:
            credentials = yaml.safe_load(file)
        print(f"Parsed YAML structure: {credentials}")

        APP_ID = credentials['twit_api']['app_id']
        APP_KEY = credentials['twit_api']['app_key']
        API_BASE_URL = credentials['twit_api'].get('base_url', 'https://twit.tv/api/v1.0')

        print(f"Using APP_ID: {APP_ID}")
        print(f"Using APP_KEY: {'*' * len(APP_KEY)}") # Mask the actual key for security
    except Exception as e:
        print(f"Error loading credentials: {e}")
        print("Please make sure your credentials.yml file is properly formatted.")
        print("You can copy credentials.yml.sample to credentials.yml and update with your actual credentials.")
        sys.exit(1)
else:
    print(f"Credentials file not found: {credentials_file}")
    print("Please copy credentials.yml.sample to credentials.yml and update with your actual credentials.")
    sys.exit(1)

# Test endpoint - shows is a simple endpoint to test
endpoint = '/shows'
url = f"{API_BASE_URL}{endpoint}"

# Create headers
headers = {
    'Accept': 'application/json',
    'app-id': APP_ID,
    'app-key': APP_KEY
}

print("\nTesting connection to TWiT API...")
print(f"Endpoint: {url}")
print(f"Headers: app-id={APP_ID}, app-key={APP_KEY}")

try:
    # Send request
    response = requests.get(url, headers=headers)

    # Parse and display response
    if response.status_code == 200:
        print(f"Connection successful! (HTTP {response.status_code})")

        # Parse JSON response
        result = response.json()

        # Display some basic info about the response
        if 'count' in result:
            print(f"Found {result['count']} shows")

        # Display first few items if available
        if 'shows' in result and result['shows']:
            print("\nFirst few shows:")
            for show in result['shows'][:3]:
                print(f"- {show['label']} ({show['id']})")
    else:
        print(f"Connection failed with HTTP status: {response.status_code}")
        print(f"Response body: {response.text}")

        # Check for specific error messages
        if response.status_code == 500 and 'usage limits are exceeded' in response.text:
            print("Error: API usage limits exceeded. Check your plan limits.")
        elif response.status_code in (401, 403):
            print("Error: Authentication failed. Check your APP_ID and APP_KEY.")
except Exception as e:
    print(f"Error connecting to API: {e}")
    import traceback
    print(traceback.format_exc())

File: examples/javascript/test_connection.js
#!/usr/bin/env node
// TWiT API Connection Test Script
// This script tests the connection to the TWiT.tv API

const https = require('https');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

// Simple YAML parser for our basic needs
function parseYAML(content) {
  const lines = content.split('\n');
  const result = {};
  let currentSection = null;
  
  for (const line of lines) {
    // Skip comments and empty lines
    if (line.trim().startsWith('#') || !line.trim()) continue;
    
    // Handle section headers (keys with no value)
    if (!line.startsWith(' ') && line.includes(':')) {
      const parts = line.split(':');
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      
      if (!value) {
        currentSection = key;
        result[currentSection] = {};
      } else {
        // Remove any inline comments from the value
        const valueParts = value.split('#');
        result[key] = valueParts[0].trim();
      }
    } 
    // Handle indented properties (section values)
    else if (currentSection && line.trim().includes(':')) {
      const parts = line.trim().split(':');
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      
      // Remove any inline comments from the value
      const valueParts = value.split('#');
      result[currentSection][key] = valueParts[0].trim();
    }
  }
  
  return result;
}

// Load credentials from shared credentials file
const scriptDir = path.dirname(__filename);
const credentialsFile = path.join(scriptDir, '..', 'credentials.yml');

if (fs.existsSync(credentialsFile)) {
  try {
    console.log(`Loading credentials from: ${credentialsFile}`);
    const credentialsContent = fs.readFileSync(credentialsFile, 'utf8');
    console.log(`Credentials file content (first line): ${credentialsContent.split('\n')[0].trim()}`);

    const credentials = parseYAML(credentialsContent);
    console.log(`Parsed YAML structure: ${JSON.stringify(credentials)}`);

    const APP_ID = credentials.twit_api.app_id;
    const APP_KEY = credentials.twit_api.app_key;
    const API_BASE_URL = credentials.twit_api.base_url || 'https://twit.tv/api/v1.0';

    console.log(`Using APP_ID: ${APP_ID}`);
    console.log(`Using APP_KEY: ${'*'.repeat(APP_KEY.length)}`); // Mask the actual key for security
    console.log(`Using BASE_URL: ${API_BASE_URL}`);

    // Test endpoint - shows is a simple endpoint to test
    const endpoint = '/shows';
    const url = `${API_BASE_URL}${endpoint}`;
    
    try {
      const parsedUrl = new URL(url);

      // Create options for request
      const options = {
        hostname: parsedUrl.hostname,
        path: parsedUrl.pathname + parsedUrl.search,
        method: 'GET',
        headers: {
          'Accept': 'application/json',
          'app-id': APP_ID,
          'app-key': APP_KEY
        }
      };

      console.log("\nTesting connection to TWiT API...");
      console.log(`Endpoint: ${url}`);
      console.log(`Headers: app-id=${APP_ID}, app-key=${APP_KEY}`);

      // Send request
      const req = https.request(options, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          // Parse and display response
          if (res.statusCode === 200) {
            console.log(`Connection successful! (HTTP ${res.statusCode})`);

            // Parse JSON response
            try {
              const result = JSON.parse(data);

              // Display some basic info about the response
              if (result.count) {
                console.log(`Found ${result.count} shows`);
              }

              // Display first few items if available
              if (result.shows && result.shows.length > 0) {
                console.log("\nFirst few shows:");
                result.shows.slice(0, 3).forEach(show => {
                  console.log(`- ${show.label} (${show.id})`);
                });
              }
            } catch (e) {
              console.error(`Error parsing JSON response: ${e.message}`);
            }
          } else {
            console.log(`Connection failed with HTTP status: ${res.statusCode}`);
            console.log(`Response body: ${data}`);

            // Check for specific error messages
            if (res.statusCode === 500 && data.includes('usage limits are exceeded')) {
              console.log("Error: API usage limits exceeded. Check your plan limits.");
            } else if (res.statusCode === 401 || res.statusCode === 403) {
              console.log("Error: Authentication failed. Check your APP_ID and APP_KEY.");
            } else if (res.statusCode === 404) {
              console.log("Error: Resource not found. Check that the API endpoint is correct.");
              console.log("The API might be at a different URL or the path might be incorrect.");
            }
          }
        });
      });

      req.on('error', (e) => {
        console.error(`Error connecting to API: ${e.message}`);
      });

      req.end();
    } catch (e) {
      console.error(`Error with URL: ${e.message}`);
      console.log(`Invalid URL: ${url}`);
      console.log("Please check the base_url in your credentials.yml file.");
    }
  } catch (e) {
    console.error(`Error loading credentials: ${e.message}`);
    console.log("Please make sure your credentials.yml file is properly formatted.");
    console.log("You can copy credentials.yml.sample to credentials.yml and update with your actual credentials.");
    process.exit(1);
  }
} else {
  console.error(`Credentials file not found: ${credentialsFile}`);
  console.log("Please copy credentials.yml.sample to credentials.yml and update with your actual credentials.");
  process.exit(1);
}
